maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -108,15 +168,31 @@ def step(self, tick: int):
         self._update_vm_workload()
         # Update all PM CPU utilization.
         self._update_pm_workload()
-        # TODO
-        # Generate VM requirement events from data file.
-        # It might be implemented by a for loop to process VMs in each tick.
-        vm_requirement_event = self._event_buffer.gen_cascade_event(
-            tick=tick,
-            event_type=Events.REQUIREMENTS,
-            payload=None
-        )
-        self._event_buffer.insert_event(event=vm_requirement_event)
+
+        self._cur_tick_cpu_utilization = self._cpu_reader.items(tick=tick)
+
+        for vm in self._vm_item_picker.items(tick):
+            vm_req = VirtualMachine(
+                id=vm.vm_id,
+                vcpu_cores_requirement=vm.vm_vcpu_cores,
+                memory_requirement=vm.vm_memory,
+                lifetime=vm.vm_deleted - vm.timestamp + 1
+            )
+            vm_req.add_utilization(cpu_utilization=self._cur_tick_cpu_utilization[vm.vm_id])
+            vm_req_payload: VmRequirementPayload = VmRequirementPayload(
+                vm_req=vm_req,
+                remaining_buffer_time=self._buffer_time_budget
+            )
+            vm_requirement_event = self._event_buffer.gen_cascade_event(
+                tick=tick,
+                event_type=Events.REQUIREMENTS,
+                payload=vm_req_payload
+            )
+            self._event_buffer.insert_event(event=vm_requirement_event)
+            self._total_vm_requirements += 1
+
+    def post_step(self, tick: int):
```

Comment:
The workflow that would effect pm is update_pm_workload -> assign action.
And it will calculate the cpu_utilization for pm when every vm is assigned to the pm.

```python
# Update PM resources requested by VM.
pm = self._machines[pm_id]
pm.place_vm(vm.id)
pm.cpu_allocation += vm.cpu_cores_requirement
pm.memory_allocation += vm.memory_requirement
# Calculate the PM's utilization.
pm_cpu_utilization = (
     (pm.cpu_cores_capacity * pm.cpu_utilization + vm.cpu_cores_requirement * vm.cpu_utilization)
      / pm.cpu_cores_capacity
)
pm.cpu_utilization = pm_cpu_utilization
pm.energy_consumption = self._cpu_utilization_to_energy_consumption(cpu_utilization=pm.cpu_utilization)
```
So I think we just need to sum the total_energy of pms and update to the matrices at the post step.

Reply:
And we don't need to calculate the energy at the `update_pm_workload` ?